clients/spark/core/src/main/scala/io/treeverse/utils/AsInterface.scala

```diff
@@ -0,0 +1,167 @@
+                                   method.getName,
+                                   Type.getMethodDescriptor(method),
+                                   method.getExceptionTypes.map(e => Type.getType(e).getDescriptor)
+                                  )
+        )
+        // Generate wrapper for all constructors for the class.
+        iface.getConstructors.foreach(ctor =>
+          generateForwardingMethod(ACC_PUBLIC | ACC_SYNTHETIC,
+                                   "<init>",
+                                   Type.getConstructorDescriptor(ctor),
+                                   ctor.getExceptionTypes.map(e => Type.getType(e).getDescriptor)
+                                  )
+        )
+      }
+      case _ => Unit
+    }
+  }
+
+  def translateType(typ: Type) =
+    ifaces get AsInterface.undot(typ.getInternalName) match {
+      case Some(iface) => Type.getType(iface)
+      case None        => typ
+    }
+
+  def revTranslateType(typ: Type) =
+    revIfaces get typ.getInternalName match {
+      case Some(klassName) => Type.getObjectType(klassName)
+      case None            => typ
+    }
+
+  def generateForwardingMethod(
+      access: Int,
+      name: String,
+      translatedDesc: String,
+      translatedExceptions: Seq[String]
+  ): Unit = {
+    val translatedArgTypes = Type.getArgumentTypes(translatedDesc)
+    if (translatedArgTypes == null) {
+      throw new IllegalArgumentException(s"Bad translated method descriptor ${translatedDesc}")
+    }
+    val argTypes = translatedArgTypes.map(revTranslateType)
+
+    val translatedRetType = Type.getReturnType(translatedDesc)
+    val retType = revTranslateType(translatedRetType)
+
+    // BUG(ariels): To handle exceptions, need to wrap INVOKE with
+    //
+    //     try {...} catch (exception) { throw translatedException }
+    val exceptions =
+      if (translatedExceptions != null)
+        translatedExceptions.map((name: String) =>
+          revTranslateType(Type.getObjectType(name)).getInternalName
+        )
+      else
+        null
+
+    val desc = Type.getMethodDescriptor(retType, argTypes: _*)
+
+    if (
+      retType.equals(translatedRetType) &&
+      (exceptions == null || AsInterface.equals(exceptions, translatedExceptions)) &&
+      AsInterface.equals(argTypes, translatedArgTypes)
+    ) {
+      return
+    }
+
+    val mv = cv.visitMethod(
+      access,
+      name,
+      translatedDesc,
+      null /* no support for generics */,
+      translatedExceptions.toArray
+    )
+
+    mv.visitCode()
+    // BUG(ariels): assumes non-static.
+    mv.visitVarInsn(ALOAD, 0) // load "this"
+
+    for (((argType, translatedArgType), index) <- (argTypes zip translatedArgTypes).zipWithIndex) {
+      mv.visitVarInsn(AsInterface.loadFor(argType.getSort), index + 1) // load arg
+      if (!argType.equals(translatedRetType)) {
+        mv.visitTypeInsn(CHECKCAST, argType.getInternalName)
+      }
+    }
```

Comment:
I don't get this. What happens if the `CHECKCAST` fail?

Reply:
An exception.  Which is exactly the exception that we want to happen here :-)